include/device.h

```diff
@@ -524,6 +564,42 @@ int device_required_foreach(const struct device *dev,
 			  device_visitor_callback_t visitor_cb,
 			  void *context);
 
+/**
+ * @brief Visit every device that @p dev directly supports.
+ *
+ * Zephyr maintains information about which devices are directly supported by
+ * another device; for example an I2C controller will support an I2C-based
+ * sensor driver. Supported devices can derive from statically-defined
+ * devicetree relationships.
+ *
+ * This API supports operating on the set of supported devices. Example uses
+ * include iterating over the devices connected to a regulator when it is
+ * powered on.
+ *
+ * There is no guarantee on the order in which required devices are visited.
+ *
+ * If the @p visitor function returns a negative value iteration is halted,
+ * and the returned value from the visitor is returned from this function.
+ *
+ * @note This API is not available to unprivileged threads.
+ *
+ * @param dev a device of interest. The devices that this device supports
+ * will be used as the set of devices to visit. This parameter must not be
+ * null.
+ *
+ * @param visitor_cb the function that should be invoked on each device in the
+ * support set. This parameter must not be null.
+ *
+ * @param context state that is passed through to the visitor function. This
+ * parameter may be null if @p visitor tolerates a null @p context.
+ *
+ * @return The number of devices that were visited if all visits succeed, or
+ * the negative value returned from the first visit that did not succeed.
+ */
+int device_supported_foreach(const struct device *dev,
```

Comment:
I am wondering if "supported" is the right language we want to use here, maybe "connected"? Then the @brief would be "Visit every device that is directly connected to @p dev."

Reply:
Supports is the terminology used in the existing API's. I do agree it makes for awkward documentation.
https://github.com/zephyrproject-rtos/zephyr/blob/cc760c54018ba142d283350e653881b11485c4ff/include/devicetree/ordinals.h#L47-L61

Reply:
> Supports is the terminology used in the existing API's. I do agree it makes for awkward documentation.

yes it looks very awkward for someone not familiar with the DT APIs. Is this a zephyr thing or does this come from the devicetree spec? @mbolivar-nordic @galak